packages/surveycto/src/Adaptor.js

```diff
@@ -98,20 +90,483 @@ export function fetchSubmissions(formId, options = {}, callback = s => s) {
 
     console.log(`Fetching '${resolvedFormId}' submissions for: ${date}`);
 
-    return requestHelper(
+    const response = await requestHelper(state, path, {
+      headers: {
+        'content-type': contentType,
+      },
+      query: {
+        date,
+        r: status,
+      },
+    });
+
+    return prepareNextState(state, response);
+  };
+}
+
+/**
+ * List datasets from the SurveyCTO API
+ * @public
+ * @example <caption>List all datasets</caption>
+ * listDatasets();
+ * @example <caption>List datasets with pagination options</caption>
+ * listDatasets({
+ *   limit: 2,
+ * });
+ * @function
+ * @param {object} options - Optional request query options. [See the API docs for details](https://developer.surveycto.com/api-v2.html#getdatasets-parameters)
+ * @property {number} options.limit - Maximum number of datasets to return. Defaults to 20. Maximum is 1000.
+ * @property {string} options.cursor - Optional string to specify the starting point of the next page of results.
+ * @returns {Operation}
+ */
+export function listDatasets(options = {}) {
+  return async state => {
+    const results = [];
+    const [resolvedOptions] = expandReferences(state, options);
+
+    const userLimit = resolvedOptions?.limit
+      ? Number(resolvedOptions.limit)
+      : undefined;
+
+    let cursor = resolvedOptions?.cursor ? resolvedOptions.cursor : null;
+
+    const baseQuery = { ...resolvedOptions };
+    delete baseQuery.limit;
+    delete baseQuery.cursor;
+
+    const pageSize = 20; // when no limit is given get 20 at a time
+    const maxFetchSize = 1000; // server max is 1000 per request
+
+    const desiredFetchTotal = userLimit ?? Infinity;
+
+    if (Number.isFinite(desiredFetchTotal) && desiredFetchTotal <= 0) {
+      return prepareNextState(state, {
+        data: [],
+        total: 0,
+        nextCursor: cursor,
+      });
+    }
+
+    do {
```

Comment:
I think all pagination stuff should be a util function eg: requestWithPagination(). It will also make it easy to write unit test and test different scenarios.